store/src/commonMain/kotlin/com/dropbox/external/store5/impl/ShareableMarket.kt

```diff
@@ -0,0 +1,431 @@
+        return try {
+            val request = getLatestWriteRequest<Input, Output>(key)
+            val last = lastOrNull<Input>(key) ?: throw Exception()
+            return when (request.post(key, last)) {
+                else -> RemoteResult.Success(last)
+            }
+        } catch (throwable: Throwable) {
+            RemoteResult.Failure(throwable)
+        }
+    }
+
+    override suspend fun delete(key: Key): Boolean {
+        for (store in stores) {
+            try {
+                when (store.delete(key)) {
+                    true -> continue
+                    false -> throw Exception()
+                }
+            } catch (throwable: Throwable) {
+                return false
+            }
+        }
+        val broadcast = requireBroadcast<Any>(key)
+        val response = MarketResponse.Empty
+        broadcast.emit(response)
+        return true
+    }
+
+    override suspend fun delete(): Boolean {
+        for (store in stores) {
+            try {
+                store.deleteAll.invoke()
+            } catch (throwable: Throwable) {
+                return false
+            }
+        }
+
+        for (broadcast in broadcasts.access { it.values }) {
+            val response = MarketResponse.Empty
+            broadcast.emit(response)
+        }
+
+        return true
+    }
+
+    private fun <Input : Any, Output : Any> getAndEmitLatest(
+        key: Key,
+        fetcher: Fetcher<Key, Input, Output>
+    ) {
+        scope.launch {
+            val isBroadcasting = async { startIfNotBroadcasting<Output>(key) }
+            isBroadcasting.await()
+            load<Output>(key)
+            refresh(key, fetcher)
+        }
+    }
+
+    private suspend fun <Output : Any> getOrSetBroadcast(key: Key): SomeBroadcast<Output> {
+
+        val result = scope.async {
+            if (broadcasts.access { it[key] == null }) {
+                broadcasts.access {
+                    it[key] = MutableSharedFlow(4)
+                    scope.launch {
+                        it[key]!!.emit(MarketResponse.Loading)
+                    }
+                }
+            }
+        }
+
+
+        result.await()
+        return requireBroadcast(key)
+    }
+
+    private fun <Output : Any> getBroadcast(key: Key): SomeBroadcast<Output>? =
+        broadcasts.access { it[key] } as? SomeBroadcast<Output>
+
+    private fun <Output : Any> requireBroadcast(key: Key): SomeBroadcast<Output> =
+        broadcasts.access { it[key]!! } as SomeBroadcast<Output>
+
+    private fun <Input : Any> requireWriteRequestQueue(key: Key): SomeWriteRequestQueue<Key, Input> =
+        writeRequests.access { it[key]!! } as SomeWriteRequestQueue<Key, Input>
+
+    private fun broadcasting(key: Key) = broadcasts.access { it[key] != null }
+
+    private fun addOrInitReadCompletions(key: Key, onCompletions: AnyReadCompletionsQueue) {
+        if (readCompletions.access { it[key] != null }) {
```

Comment:
I don't think these are thread safe. I think you want to perform a single block in a single `access` class.

Reply:
Take another look at `RealMarket` on Mon please. Going to resolve these as outdated